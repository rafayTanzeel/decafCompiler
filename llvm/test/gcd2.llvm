; ModuleID = 'GreatestCommonDivisor'

@a = internal global i32 10
@b = internal global i32 20

declare void @print_int(i32)

define i32 @main() {
entry:
  %z = alloca i32
  %a = load i32, i32* @a
  %b = load i32, i32* @b
  %calltmp = call i32 @gcd(i32 %a, i32 %b)
  call void @print_int(i32 %calltmp)
  ret i32 0
}

define i1 @iszero(i32 %a, i32 %b) {
entry:
  %a1 = alloca i32
  store i32 %a, i32* %a1
  %b2 = alloca i32
  store i32 %b, i32* %b2
  br label %ifstart
}

define i32 @gcd(i32 %a, i32 %b) {
entry:
  %a1 = alloca i32
  store i32 %a, i32* %a1
  %b2 = alloca i32
  store i32 %b, i32* %b2
  %a8 = load i32, i32* %a1
  %b9 = load i32, i32* %b2
  %calltmp10 = call i1 @iszero(i32 %a8, i32 %b9)
  ret i1 %calltmp10

ifstart:                                          ; preds = %entry
  %b3 = load i32, i32* %b2
  %ICmpEQtmp = icmp eq i32 %b3, 0
  br i1 %ICmpEQtmp, label %iftrue, label %iffalse

iftrue:                                           ; preds = %ifstart
  %a4 = load i32, i32* %a1
  ret i32 %a4
  br label %end

end:                                              ; preds = %iffalse, %iftrue
  ret i32 0

iffalse:                                          ; preds = %ifstart
  %b5 = load i32, i32* %b2
  %b6 = load i32, i32* %b2
  %a7 = load i32, i32* %a1
  %sremtmp = srem i32 %a7, %b6
  %calltmp = call i32 @gcd(i32 %b5, i32 %sremtmp)
  ret i32 %calltmp
  br label %end
}
