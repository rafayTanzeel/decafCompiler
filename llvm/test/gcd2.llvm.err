; ModuleID = 'GreatestCommonDivisor'

@a = internal global i32 10
@b = internal global i32 20

declare void @print_int(i32)

define i32 @main() {
entry:
  %z = alloca i32
  %a = load i32, i32* @a
  %b = load i32, i32* @b
  %calltmp = call i32 @gcd(i32 %a, i32 %b)
  call void @print_int(i32 %calltmp)
  ret i32 0
}

define i32 @gcd(i32 %a, i32 %b) {
entry:
  %a1 = alloca i32
  store i32 %a, i32* %a1
  %b2 = alloca i32
  store i32 %b, i32* %b2
  %a3 = load i32, i32* %a1
  %b4 = load i32, i32* %b2
  %calltmp = call i1 @iszero(i32 %a3, i32 %b4)
  %zexttmp = zext i1 %calltmp to i32
  ret i32 %zexttmp
}

define i1 @iszero(i32 %a, i32 %b) {
entry:
  %a1 = alloca i32
  store i32 %a, i32* %a1
  %b2 = alloca i32
  store i32 %b, i32* %b2
  br label %ifstart

ifstart:                                          ; preds = %entry
  %b3 = load i32, i32* %b2
  %ICmpEQtmp = icmp eq i32 %b3, 0
  br i1 %ICmpEQtmp, label %iftrue, label %iffalse

iftrue:                                           ; preds = %ifstart
  %a4 = load i32, i32* %a1
  ret i1 true
  br label %end

end:                                              ; preds = %iffalse, %iftrue
  ret i1 true

iffalse:                                          ; preds = %ifstart
  %b5 = load i32, i32* %b2
  %a6 = load i32, i32* %a1
  %b7 = load i32, i32* %b2
  %sremtmp = srem i32 %a6, %b7
  %calltmp = call i32 @gcd(i32 %b5, i32 %sremtmp)
  ret i1 true
  br label %end
}
