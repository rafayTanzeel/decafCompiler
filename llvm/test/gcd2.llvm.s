	.text
	.file	"llvm/test/gcd2.llvm.bc"
	.globl	main
	.align	16, 0x90
	.type	main,@function
main:                                   # @main
	.cfi_startproc
# BB#0:                                 # %entry
	pushq	%rax
.Ltmp0:
	.cfi_def_cfa_offset 16
	movl	a(%rip), %edi
	movl	b(%rip), %esi
	callq	gcd
	movl	%eax, %edi
	callq	print_int
	xorl	%eax, %eax
	popq	%rcx
	retq
.Lfunc_end0:
	.size	main, .Lfunc_end0-main
	.cfi_endproc

	.globl	gcd
	.align	16, 0x90
	.type	gcd,@function
gcd:                                    # @gcd
	.cfi_startproc
# BB#0:                                 # %entry
	pushq	%rax
.Ltmp1:
	.cfi_def_cfa_offset 16
	movl	%edi, 4(%rsp)
	movl	%esi, (%rsp)
	movl	4(%rsp), %edi
	callq	iszero
	andb	$1, %al
	movzbl	%al, %eax
	popq	%rcx
	retq
.Lfunc_end1:
	.size	gcd, .Lfunc_end1-gcd
	.cfi_endproc

	.globl	iszero
	.align	16, 0x90
	.type	iszero,@function
iszero:                                 # @iszero
	.cfi_startproc
# BB#0:                                 # %entry
	pushq	%rax
.Ltmp2:
	.cfi_def_cfa_offset 16
	movl	%edi, 4(%rsp)
	movl	%esi, (%rsp)
	cmpl	$0, (%rsp)
	je	.LBB2_2
# BB#1:                                 # %iffalse
	movl	(%rsp), %edi
	movl	4(%rsp), %eax
	cltd
	idivl	%edi
	movl	%edx, %esi
	callq	gcd
.LBB2_2:                                # %iftrue
	movb	$1, %al
	popq	%rcx
	retq
.Lfunc_end2:
	.size	iszero, .Lfunc_end2-iszero
	.cfi_endproc

	.type	a,@object               # @a
	.data
	.align	4
a:
	.long	10                      # 0xa
	.size	a, 4

	.type	b,@object               # @b
	.align	4
b:
	.long	20                      # 0x14
	.size	b, 4


	.section	".note.GNU-stack","",@progbits
