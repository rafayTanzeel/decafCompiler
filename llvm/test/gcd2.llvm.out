semantic error: function not found
